lib/Infer/ExhaustiveSynthesis.cpp

```diff
@@ -86,6 +88,60 @@ namespace {
 // experiment with synthesizing at reduced bitwidth, then expanding the result
 // aggressively avoid calling into the solver
 
+using CostFunctionType = std::function<float(Inst *, bool)>;
+
+float defaultCostFunction(Inst* I, bool IgnoreDepsWithExternalUses) {
+  return souper::cost(I, IgnoreDepsWithExternalUses);
+}
+
+// TODO : Derive these randomly to maximize infer rate with test-infer.sh
+float getCost(Inst::Kind K) {
+  switch (K) {
+    case souper::Inst::Var:
+    case souper::Inst::Const:
+    case souper::Inst::Phi:
+      return 0;
+    case souper::Inst::BSwap:
+    case souper::Inst::CtPop:
+    case souper::Inst::Cttz:
+    case souper::Inst::Ctlz:
+    case souper::Inst::SDiv:
+    case souper::Inst::UDiv:
+    case souper::Inst::SRem:
+    case souper::Inst::URem:
+      return 5;
+    case souper::Inst::Select:
+      return 3;
+    case souper::Inst::Mul:
+      return 2.5f;
+    case souper::Inst::Add:
+    case souper::Inst::Sub:
+      return 1.5f;
+    default:
+      return 1;
+  }
+}
+
+// TODO: Recognize simple patterns
+static float costHelper(Inst *I, Inst *Root, std::set<Inst *> &Visited,
+                      bool IgnoreDepsWithExternalUses) {
+  if (!Visited.insert(I).second)
+    return 0;
+  if (IgnoreDepsWithExternalUses && I != Root &&
```

Comment:
I think you should leave a comment explaining the rationale for counting things with external uses are zero cost

Reply:
> I agree that we want pluggable cost functions, but do you need to add a new one right now? it is very similar to the existing one, so maybe the existing one simply wants to be replaced (while also leaving the pluggable infrastructure so that we can plug in more sophisticated stuff later)?

I think it makes sense to keep the existing one because the CEGIS synthesis needs it.
Also, I want to try adding a few variants and maybe a parametric one that can be controlled from the command line.